tss-esapi-sys/build.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright 2021 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+const MINIMUM_VERSION: &str = "2.3.3";
+
+fn main() {
+    #[cfg(feature = "generate-bindings")]
+    {
+        generate::generate_bindings();
+    }
+
+    #[cfg(not(feature = "generate-bindings"))]
+    {
+        let supported_platforms = vec![String::from("x86_64-unknown-linux-gnu")];
+        let target = std::env::var("TARGET").unwrap();
+
+        // check if target is in the list of supported ones or panic with nice message
+        if !supported_platforms.contains(&target) {
+            panic!(format!("Compilation target ({}) is not part of the supported targets ({:?}). Please compile with the \"generate-bindings\" feature or add support for your platform :)", target, supported_platforms));
+        }
+
+        pkg_config::Config::new()
+            .atleast_version(MINIMUM_VERSION)
+            .probe("tss2-esys")
+            .expect("Failed to find tss2-esys library.");
+        pkg_config::Config::new()
+            .atleast_version(MINIMUM_VERSION)
+            .probe("tss2-tctildr")
+            .expect("Failed to find tss2-tctildr library.");
+        pkg_config::Config::new()
+            .atleast_version(MINIMUM_VERSION)
+            .probe("tss2-mu")
+            .expect("Failed to find tss2-mu library.");
+    }
+}
+
+#[cfg(feature = "generate-bindings")]
+mod generate {
+    const TSS_INCLUDE_PATH: &str = "vendor/include/tss2";
+
+    use super::MINIMUM_VERSION;
+    use std::path::PathBuf;
+
+    // Only on a specific feature
+    pub fn generate_bindings() {
+        let out_path = std::path::PathBuf::from(std::env::var("OUT_DIR").unwrap());
+        let esys_path = out_path.join("tss_esapi_bindings.rs");
+        generate_from_system(esys_path.clone())
```

Comment:
I'd prefer it if `generate_from_submodule` was a separate feature, instead of the fallback if generating from the system fails.
Because on Fedora, we'd always want (need) to do the system generation, and if that fails, we'd want to know that that failed and why.

Reply:
I was wondering actually, should we have the `generate_from_submodule` fallback? I believe it is only used for the headers and not for compilation? How would someone then link to its TSS libraries when using the `generate_from_submodule` fallback?

If we give the constraint that `pkg-config` must be present on the system that that it must locate successfully the TSS libraries then we can simplify:
* by default, the crate will use its pre-generated bindings
* when `generate-bindings` is activated, we use `pkg-config` to find the headers and generate them

We would not need the `vendor` submodule then.

In the future, we could think about other modes:
1. The libraries are compiled by the crate itself to be linked locally and the bindings generated (or not)
2. The path to a custom location of the header files and libraries is given to generate the bindings (or not) and link (through environment variables or dynamically load at runtime)

But I think that, at least for now, the first two options should be enough in most use-cases?

Reply:
Yeah, make sense, I've removed the submodule - keeping just generation from the system headers